File: solutions/0049. Group Anagrams/0049.py
# Time complexity: O(nklogk)
# Space complexity: O(nk)
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        dict = collections.defaultdict(list)
        
        for str in strs:
            key = ''.join(sorted(str))
            dict[key].append(str)

        return dict.values() 

File: solutions/0049. Group Anagrams/0049.cpp
// Time complexity: O(nklogk)
// Space complexity: O(nk)
class Solution {
    vector<vector<string>> ans;
    unordered_map<string, vector<string>> keyToAnagrams;

    for (const string& str : strs) {
        string key = str;
        ranges::sort(key);
        keyToAnagrams[key].push_back(str);
    }

    for (const auto& [_, anagrams] : keyToAnagrams)
        ans.push_back(anagrams);

    return ans;    
};